// @activepieces/shared/src/lib/flows/flow-instances/flow-instance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowInstance = exports.FlowScheduleOptions = exports.FlowInstanceStatus = exports.ScheduleType = void 0;
const typebox_1 = require("@sinclair/typebox");
const common_1 = require("../../common");
var ScheduleType;
(function (ScheduleType) {
    ScheduleType["CRON_EXPRESSION"] = "CRON_EXPRESSION";
})(ScheduleType || (exports.ScheduleType = ScheduleType = {}));
var FlowInstanceStatus;
(function (FlowInstanceStatus) {
    FlowInstanceStatus["ENABLED"] = "ENABLED";
    FlowInstanceStatus["DISABLED"] = "DISABLED";
    FlowInstanceStatus["UNPUBLISHED"] = "UNPUBLISHED";
})(FlowInstanceStatus || (exports.FlowInstanceStatus = FlowInstanceStatus = {}));
exports.FlowScheduleOptions = typebox_1.Type.Optional(typebox_1.Type.Object({
    type: typebox_1.Type.Literal(ScheduleType.CRON_EXPRESSION),
    cronExpression: typebox_1.Type.String(),
    timezone: typebox_1.Type.String(),
}));
exports.FlowInstance = typebox_1.Type.Object(Object.assign(Object.assign({}, common_1.BaseModelSchema), { projectId: typebox_1.Type.String(), flowId: typebox_1.Type.String(), flowVersionId: typebox_1.Type.String(), schedule: exports.FlowScheduleOptions, status: typebox_1.Type.Enum(FlowInstanceStatus) }));
