// @activepieces/shared/src/lib/analytics/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnalyticsReportResponse = exports.AnalyticsProjectReport = exports.AnalyticsProjectReportItem = exports.AnalyticsPieceReport = exports.AnalyticsPieceReportItem = void 0;
const typebox_1 = require("@sinclair/typebox");
exports.AnalyticsPieceReportItem = typebox_1.Type.Object({
    name: typebox_1.Type.String(),
    displayName: typebox_1.Type.String(),
    logoUrl: typebox_1.Type.String(),
    usageCount: typebox_1.Type.Number(),
});
exports.AnalyticsPieceReport = typebox_1.Type.Array(exports.AnalyticsPieceReportItem);
exports.AnalyticsProjectReportItem = typebox_1.Type.Object({
    id: typebox_1.Type.String(),
    displayName: typebox_1.Type.String(),
    activeFlows: typebox_1.Type.Number(),
    totalFlows: typebox_1.Type.Number(),
});
exports.AnalyticsProjectReport = typebox_1.Type.Array(exports.AnalyticsProjectReportItem);
exports.AnalyticsReportResponse = typebox_1.Type.Object({
    totalFlows: typebox_1.Type.Number(),
    activeFlows: typebox_1.Type.Number(),
    totalUsers: typebox_1.Type.Number(),
    activeUsers: typebox_1.Type.Number(),
    totalProjects: typebox_1.Type.Number(),
    activeProjects: typebox_1.Type.Number(),
    uniquePiecesUsed: typebox_1.Type.Number(),
    activeFlowsWithAI: typebox_1.Type.Number(),
    topPieces: exports.AnalyticsPieceReport,
    tasksUsage: typebox_1.Type.Array(typebox_1.Type.Object({
        day: typebox_1.Type.String(),
        totalTasks: typebox_1.Type.Number(),
    })),
    topProjects: exports.AnalyticsProjectReport,
});
