// @activepieces/shared/src/lib/flows/flow-operations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowOperationRequest = exports.UpdateTriggerRequest = exports.AddActionRequest = exports.MoveActionRequest = exports.UpdateActionRequest = exports.DeleteActionRequest = exports.ChangeNameRequest = exports.ChangeFolderRequest = exports.ImportFlowRequest = exports.LockFlowRequest = exports.StepLocationRelativeToParent = exports.FlowOperationType = void 0;
const action_1 = require("./actions/action");
const trigger_1 = require("./triggers/trigger");
const typebox_1 = require("@sinclair/typebox");
var FlowOperationType;
(function (FlowOperationType) {
    FlowOperationType["LOCK_FLOW"] = "LOCK_FLOW";
    FlowOperationType["CHANGE_FOLDER"] = "CHANGE_FOLDER";
    FlowOperationType["CHANGE_NAME"] = "CHANGE_NAME";
    FlowOperationType["MOVE_ACTION"] = "MOVE_ACTION";
    FlowOperationType["IMPORT_FLOW"] = "IMPORT_FLOW";
    FlowOperationType["UPDATE_TRIGGER"] = "UPDATE_TRIGGER";
    FlowOperationType["ADD_ACTION"] = "ADD_ACTION";
    FlowOperationType["UPDATE_ACTION"] = "UPDATE_ACTION";
    FlowOperationType["DELETE_ACTION"] = "DELETE_ACTION";
})(FlowOperationType || (exports.FlowOperationType = FlowOperationType = {}));
var StepLocationRelativeToParent;
(function (StepLocationRelativeToParent) {
    StepLocationRelativeToParent["INSIDE_TRUE_BRANCH"] = "INSIDE_TRUE_BRANCH";
    StepLocationRelativeToParent["INSIDE_FALSE_BRANCH"] = "INSIDE_FALSE_BRANCH";
    StepLocationRelativeToParent["AFTER"] = "AFTER";
    StepLocationRelativeToParent["INSIDE_LOOP"] = "INSIDE_LOOP";
})(StepLocationRelativeToParent || (exports.StepLocationRelativeToParent = StepLocationRelativeToParent = {}));
const optionalNextAction = typebox_1.Type.Object({ nextAction: typebox_1.Type.Optional(action_1.Action) });
exports.LockFlowRequest = typebox_1.Type.Object({
    flowId: typebox_1.Type.String({})
});
exports.ImportFlowRequest = typebox_1.Type.Object({
    displayName: typebox_1.Type.String({}),
    trigger: typebox_1.Type.Union([typebox_1.Type.Composite([trigger_1.WebhookTrigger, optionalNextAction]), typebox_1.Type.Composite([trigger_1.PieceTrigger, optionalNextAction]), typebox_1.Type.Composite([trigger_1.EmptyTrigger, optionalNextAction])]),
});
exports.ChangeFolderRequest = typebox_1.Type.Object({
    folderId: typebox_1.Type.Union([typebox_1.Type.String(), typebox_1.Type.Null()])
});
exports.ChangeNameRequest = typebox_1.Type.Object({
    displayName: typebox_1.Type.String({}),
});
exports.DeleteActionRequest = typebox_1.Type.Object({
    name: typebox_1.Type.String()
});
exports.UpdateActionRequest = typebox_1.Type.Union([action_1.CodeActionSchema, action_1.LoopOnItemsActionSchema, action_1.PieceActionSchema, action_1.BranchActionSchema, action_1.MissingActionSchema]);
exports.MoveActionRequest = typebox_1.Type.Object({
    name: typebox_1.Type.String(),
    newParentStep: typebox_1.Type.String(),
    stepLocationRelativeToNewParent: typebox_1.Type.Optional(typebox_1.Type.Enum(StepLocationRelativeToParent))
});
exports.AddActionRequest = typebox_1.Type.Object({
    parentStep: typebox_1.Type.String(),
    stepLocationRelativeToParent: typebox_1.Type.Optional(typebox_1.Type.Enum(StepLocationRelativeToParent)),
    action: exports.UpdateActionRequest
});
exports.UpdateTriggerRequest = typebox_1.Type.Union([trigger_1.EmptyTrigger, trigger_1.PieceTrigger, trigger_1.WebhookTrigger]);
exports.FlowOperationRequest = typebox_1.Type.Union([
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(FlowOperationType.MOVE_ACTION),
        request: exports.MoveActionRequest
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(FlowOperationType.LOCK_FLOW),
        request: exports.LockFlowRequest
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(FlowOperationType.IMPORT_FLOW),
        request: exports.ImportFlowRequest
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(FlowOperationType.CHANGE_NAME),
        request: exports.ChangeNameRequest
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(FlowOperationType.DELETE_ACTION),
        request: exports.DeleteActionRequest
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(FlowOperationType.UPDATE_ACTION),
        request: exports.UpdateActionRequest
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(FlowOperationType.ADD_ACTION),
        request: exports.AddActionRequest
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(FlowOperationType.UPDATE_TRIGGER),
        request: exports.UpdateTriggerRequest
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(FlowOperationType.CHANGE_FOLDER),
        request: exports.ChangeFolderRequest
    })
]);
