// @activepieces/shared/src/lib/manual-task/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManualTaskWithAssignee = exports.ManualTask = exports.StatusOption = exports.STATUS_COLORS = exports.RESOLVED_STATUS = exports.UNRESOLVED_STATUS = exports.STATUS_VARIANT = void 0;
const typebox_1 = require("@sinclair/typebox");
const common_1 = require("../common");
const user_1 = require("../user");
var STATUS_VARIANT;
(function (STATUS_VARIANT) {
    STATUS_VARIANT["POSITIVE"] = "Postive (Green)";
    STATUS_VARIANT["NEGATIVE"] = "Negative (Red)";
    STATUS_VARIANT["NEUTRAL"] = "Neutral (Gray)";
})(STATUS_VARIANT || (exports.STATUS_VARIANT = STATUS_VARIANT = {}));
exports.UNRESOLVED_STATUS = {
    name: 'Unresolved',
    description: 'Unresolved',
    variant: STATUS_VARIANT.NEUTRAL,
};
exports.RESOLVED_STATUS = {
    name: 'Resolved',
    description: 'Resolved',
    variant: STATUS_VARIANT.POSITIVE,
};
exports.STATUS_COLORS = {
    [STATUS_VARIANT.POSITIVE]: {
        color: '#e5efe7',
        textColor: '#28813e',
    },
    [STATUS_VARIANT.NEGATIVE]: {
        color: '#fbe2e3',
        textColor: '#dd111b',
    },
    [STATUS_VARIANT.NEUTRAL]: {
        color: '#f6f6f6',
        textColor: '#2c2c2c',
    },
};
exports.StatusOption = typebox_1.Type.Object({
    name: typebox_1.Type.String(),
    description: (0, common_1.Nullable)(typebox_1.Type.String()),
    variant: typebox_1.Type.Union([typebox_1.Type.Literal(STATUS_VARIANT.POSITIVE), typebox_1.Type.Literal(STATUS_VARIANT.NEGATIVE), typebox_1.Type.Literal(STATUS_VARIANT.NEUTRAL)]),
});
exports.ManualTask = typebox_1.Type.Object(Object.assign(Object.assign({}, common_1.BaseModelSchema), { title: typebox_1.Type.String(), description: (0, common_1.Nullable)(typebox_1.Type.String()), status: exports.StatusOption, statusOptions: typebox_1.Type.Array(exports.StatusOption), platformId: typebox_1.Type.String(), projectId: typebox_1.Type.String(), flowId: typebox_1.Type.String(), runId: typebox_1.Type.String(), assigneeId: (0, common_1.Nullable)(typebox_1.Type.String()), approvalUrl: (0, common_1.Nullable)(typebox_1.Type.String()) }));
exports.ManualTaskWithAssignee = typebox_1.Type.Composite([exports.ManualTask, typebox_1.Type.Object({
        assignee: (0, common_1.Nullable)(user_1.UserWithMetaInformation),
    })]);
