// @activepieces/shared/src/lib/manual-task/manual-task.request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateManualTaskRequestBody = exports.UpdateManualTaskRequestBody = exports.ListManualTaskAssigneesRequestQuery = exports.ListManualTasksQueryParams = void 0;
const typebox_1 = require("@sinclair/typebox");
const id_generator_1 = require("../common/id-generator");
const _1 = require(".");
const StatusOptionsSchema = typebox_1.Type.Array(_1.StatusOption, { minItems: 1 });
exports.ListManualTasksQueryParams = typebox_1.Type.Object({
    platformId: id_generator_1.ApId,
    projectId: id_generator_1.ApId,
    flowId: typebox_1.Type.Optional(id_generator_1.ApId),
    cursor: typebox_1.Type.Optional(typebox_1.Type.String()),
    limit: typebox_1.Type.Optional(typebox_1.Type.Integer({ minimum: 1, maximum: 100 })),
    assigneeId: typebox_1.Type.Optional(id_generator_1.ApId),
    statusOptions: typebox_1.Type.Optional(typebox_1.Type.Array(typebox_1.Type.String())),
    title: typebox_1.Type.Optional(typebox_1.Type.String()),
});
exports.ListManualTaskAssigneesRequestQuery = typebox_1.Type.Object({});
exports.UpdateManualTaskRequestBody = typebox_1.Type.Object({
    title: typebox_1.Type.Optional(typebox_1.Type.String()),
    description: typebox_1.Type.Optional(typebox_1.Type.String()),
    status: typebox_1.Type.Optional(_1.StatusOption),
    statusOptions: typebox_1.Type.Optional(StatusOptionsSchema),
    assigneeId: typebox_1.Type.Optional(id_generator_1.ApId),
});
exports.CreateManualTaskRequestBody = typebox_1.Type.Object({
    title: typebox_1.Type.String(),
    description: typebox_1.Type.Optional(typebox_1.Type.String()),
    statusOptions: StatusOptionsSchema,
    flowId: id_generator_1.ApId,
    runId: typebox_1.Type.Optional(id_generator_1.ApId),
    assigneeId: typebox_1.Type.Optional(id_generator_1.ApId),
    approvalUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
});
