// @activepieces/shared/src/lib/project-release/project-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectSyncPlan = exports.ProjectSyncPlanOperation = exports.ProjectSyncError = exports.DiffState = exports.ConnectionOperation = exports.ProjectOperation = exports.ProjectState = exports.ConnectionState = exports.FlowState = exports.ConnectionOperationType = exports.ProjectOperationType = void 0;
const typebox_1 = require("@sinclair/typebox");
const flow_1 = require("../flows/flow");
var ProjectOperationType;
(function (ProjectOperationType) {
    ProjectOperationType["UPDATE_FLOW"] = "UPDATE_FLOW";
    ProjectOperationType["CREATE_FLOW"] = "CREATE_FLOW";
    ProjectOperationType["DELETE_FLOW"] = "DELETE_FLOW";
})(ProjectOperationType || (exports.ProjectOperationType = ProjectOperationType = {}));
var ConnectionOperationType;
(function (ConnectionOperationType) {
    ConnectionOperationType["UPDATE_CONNECTION"] = "UPDATE_CONNECTION";
    ConnectionOperationType["CREATE_CONNECTION"] = "CREATE_CONNECTION";
})(ConnectionOperationType || (exports.ConnectionOperationType = ConnectionOperationType = {}));
exports.FlowState = typebox_1.Type.Omit(flow_1.PopulatedFlow, ['externalId']);
exports.ConnectionState = typebox_1.Type.Object({
    externalId: typebox_1.Type.String(),
    pieceName: typebox_1.Type.String(),
    displayName: typebox_1.Type.String(),
});
exports.ProjectState = typebox_1.Type.Object({
    flows: typebox_1.Type.Array(flow_1.PopulatedFlow),
    // NOTE: This is optional because in old releases, the connections state is not present
    connections: typebox_1.Type.Optional(typebox_1.Type.Array(exports.ConnectionState)),
});
exports.ProjectOperation = typebox_1.Type.Union([
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(ProjectOperationType.UPDATE_FLOW),
        newFlowState: exports.FlowState,
        flowState: exports.FlowState,
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(ProjectOperationType.CREATE_FLOW),
        flowState: exports.FlowState,
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(ProjectOperationType.DELETE_FLOW),
        flowState: exports.FlowState,
    }),
]);
exports.ConnectionOperation = typebox_1.Type.Union([
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(ConnectionOperationType.UPDATE_CONNECTION),
        newConnectionState: exports.ConnectionState,
        connectionState: exports.ConnectionState,
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(ConnectionOperationType.CREATE_CONNECTION),
        connectionState: exports.ConnectionState,
    }),
]);
exports.DiffState = typebox_1.Type.Object({
    operations: typebox_1.Type.Array(exports.ProjectOperation),
    connections: typebox_1.Type.Array(exports.ConnectionOperation),
});
exports.ProjectSyncError = typebox_1.Type.Object({
    flowId: typebox_1.Type.String(),
    message: typebox_1.Type.String(),
});
exports.ProjectSyncPlanOperation = typebox_1.Type.Union([
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(ProjectOperationType.CREATE_FLOW),
        flow: typebox_1.Type.Object({
            id: typebox_1.Type.String(),
            displayName: typebox_1.Type.String(),
        }),
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(ProjectOperationType.UPDATE_FLOW),
        flow: typebox_1.Type.Object({
            id: typebox_1.Type.String(),
            displayName: typebox_1.Type.String(),
        }),
        targetFlow: typebox_1.Type.Object({
            id: typebox_1.Type.String(),
            displayName: typebox_1.Type.String(),
        }),
    }),
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal(ProjectOperationType.DELETE_FLOW),
        flow: typebox_1.Type.Object({
            id: typebox_1.Type.String(),
            displayName: typebox_1.Type.String(),
        }),
    }),
]);
exports.ProjectSyncPlan = typebox_1.Type.Object({
    operations: typebox_1.Type.Array(exports.ProjectSyncPlanOperation),
    connections: typebox_1.Type.Array(exports.ConnectionOperation),
    errors: typebox_1.Type.Array(exports.ProjectSyncError),
});
