// @activepieces/shared/src/lib/platform/platform.request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminAddPlatformRequestBody = exports.UpdatePlatformRequestBody = void 0;
const typebox_1 = require("@sinclair/typebox");
const common_1 = require("../common");
const id_generator_1 = require("../common/id-generator");
const federated_authn_1 = require("../federated-authn");
const platform_model_1 = require("./platform.model");
exports.UpdatePlatformRequestBody = typebox_1.Type.Object({
    name: typebox_1.Type.Optional(typebox_1.Type.String({
        pattern: common_1.SAFE_STRING_PATTERN,
    })),
    primaryColor: typebox_1.Type.Optional(typebox_1.Type.String()),
    logoIconUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
    fullLogoUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
    favIconUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
    filteredPieceNames: typebox_1.Type.Optional(typebox_1.Type.Array(typebox_1.Type.String())),
    filteredPieceBehavior: typebox_1.Type.Optional(typebox_1.Type.Enum(platform_model_1.FilteredPieceBehavior)),
    smtp: typebox_1.Type.Optional(typebox_1.Type.Union([platform_model_1.SMTPInformation, typebox_1.Type.Null()])),
    federatedAuthProviders: typebox_1.Type.Optional(federated_authn_1.FederatedAuthnProviderConfig),
    cloudAuthEnabled: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    emailAuthEnabled: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    allowedAuthDomains: typebox_1.Type.Optional(typebox_1.Type.Array(typebox_1.Type.String())),
    enforceAllowedAuthDomains: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    pinnedPieces: typebox_1.Type.Optional(typebox_1.Type.Array(typebox_1.Type.String())),
    defaultLocale: typebox_1.Type.Optional(typebox_1.Type.Enum(common_1.LocalesEnum)),
    copilotSettings: typebox_1.Type.Optional(platform_model_1.CopilotSettings),
});
exports.AdminAddPlatformRequestBody = typebox_1.Type.Object({
    userId: id_generator_1.ApId,
    projectId: id_generator_1.ApId,
    name: typebox_1.Type.String(),
    domain: typebox_1.Type.String(),
});
