// @activepieces/shared/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApId = exports.apId = exports.ExecutionState = exports.RunEnvironment = exports.Flow = exports.FlowVersionState = exports.FlowVersion = exports.AUTHENTICATION_PROPERTY_NAME = exports.TriggerType = exports.WebhookTrigger = exports.PieceTrigger = exports.PieceTriggerSettings = exports.EmptyTrigger = exports.Trigger = exports.TestFlowRunRequestBody = exports.CodeActionSettings = exports.ActionType = exports.Action = exports.LoopOnItemsActionSettings = exports.PieceActionSettings = exports.PrincipalType = exports.SignInRequest = exports.SignUpRequest = exports.ExecuteCodeRequest = exports.StepOutputStatus = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./lib/flows/actions/action"), exports);
tslib_1.__exportStar(require("./lib/app-connection/app-connection"), exports);
tslib_1.__exportStar(require("./lib/app-connection/dto/read-app-connection-request"), exports);
tslib_1.__exportStar(require("./lib/app-connection/dto/upsert-app-connection-request"), exports);
tslib_1.__exportStar(require("./lib/common"), exports);
tslib_1.__exportStar(require("./lib/common/activepieces-error"), exports);
tslib_1.__exportStar(require("./lib/common/telemetry"), exports);
tslib_1.__exportStar(require("./lib/engine/engine-operation"), exports);
tslib_1.__exportStar(require("./lib/flag/flag"), exports);
tslib_1.__exportStar(require("./lib/flow-run/dto/list-flow-runs-request"), exports);
tslib_1.__exportStar(require("./lib/flow-run/execution/execution-output"), exports);
tslib_1.__exportStar(require("./lib/flow-run/execution/execution-type"), exports);
tslib_1.__exportStar(require("./lib/flow-run/execution/step-output"), exports);
tslib_1.__exportStar(require("./lib/flows/flow-operations"), exports);
tslib_1.__exportStar(require("./lib/flows/step-run"), exports);
tslib_1.__exportStar(require("./lib/app-connection/app-connection"), exports);
tslib_1.__exportStar(require("./lib/app-connection/dto/upsert-app-connection-request"), exports);
tslib_1.__exportStar(require("./lib/flow-run/execution/execution-output"), exports);
var step_output_1 = require("./lib/flow-run/execution/step-output");
Object.defineProperty(exports, "StepOutputStatus", { enumerable: true, get: function () { return step_output_1.StepOutputStatus; } });
tslib_1.__exportStar(require("./lib/pieces"), exports);
tslib_1.__exportStar(require("./lib/store-entry/dto/store-entry-request"), exports);
tslib_1.__exportStar(require("./lib/webhook"), exports);
tslib_1.__exportStar(require("./lib/flows/dto/count-flows-request"), exports);
var code_request_1 = require("./lib/code/dto/code-request");
Object.defineProperty(exports, "ExecuteCodeRequest", { enumerable: true, get: function () { return code_request_1.ExecuteCodeRequest; } });
var sign_up_request_1 = require("./lib/authentication/dto/sign-up-request");
Object.defineProperty(exports, "SignUpRequest", { enumerable: true, get: function () { return sign_up_request_1.SignUpRequest; } });
var sign_in_request_1 = require("./lib/authentication/dto/sign-in-request");
Object.defineProperty(exports, "SignInRequest", { enumerable: true, get: function () { return sign_in_request_1.SignInRequest; } });
var principal_type_1 = require("./lib/authentication/model/principal-type");
Object.defineProperty(exports, "PrincipalType", { enumerable: true, get: function () { return principal_type_1.PrincipalType; } });
var action_1 = require("./lib/flows/actions/action");
Object.defineProperty(exports, "PieceActionSettings", { enumerable: true, get: function () { return action_1.PieceActionSettings; } });
Object.defineProperty(exports, "LoopOnItemsActionSettings", { enumerable: true, get: function () { return action_1.LoopOnItemsActionSettings; } });
Object.defineProperty(exports, "Action", { enumerable: true, get: function () { return action_1.Action; } });
Object.defineProperty(exports, "ActionType", { enumerable: true, get: function () { return action_1.ActionType; } });
Object.defineProperty(exports, "CodeActionSettings", { enumerable: true, get: function () { return action_1.CodeActionSettings; } });
tslib_1.__exportStar(require("./lib/user/user"), exports);
var test_flow_run_request_1 = require("./lib/flow-run/test-flow-run-request");
Object.defineProperty(exports, "TestFlowRunRequestBody", { enumerable: true, get: function () { return test_flow_run_request_1.TestFlowRunRequestBody; } });
var trigger_1 = require("./lib/flows/triggers/trigger");
Object.defineProperty(exports, "Trigger", { enumerable: true, get: function () { return trigger_1.Trigger; } });
Object.defineProperty(exports, "EmptyTrigger", { enumerable: true, get: function () { return trigger_1.EmptyTrigger; } });
Object.defineProperty(exports, "PieceTriggerSettings", { enumerable: true, get: function () { return trigger_1.PieceTriggerSettings; } });
Object.defineProperty(exports, "PieceTrigger", { enumerable: true, get: function () { return trigger_1.PieceTrigger; } });
Object.defineProperty(exports, "WebhookTrigger", { enumerable: true, get: function () { return trigger_1.WebhookTrigger; } });
Object.defineProperty(exports, "TriggerType", { enumerable: true, get: function () { return trigger_1.TriggerType; } });
Object.defineProperty(exports, "AUTHENTICATION_PROPERTY_NAME", { enumerable: true, get: function () { return trigger_1.AUTHENTICATION_PROPERTY_NAME; } });
var flow_version_1 = require("./lib/flows/flow-version");
Object.defineProperty(exports, "FlowVersion", { enumerable: true, get: function () { return flow_version_1.FlowVersion; } });
Object.defineProperty(exports, "FlowVersionState", { enumerable: true, get: function () { return flow_version_1.FlowVersionState; } });
var flow_1 = require("./lib/flows/flow");
Object.defineProperty(exports, "Flow", { enumerable: true, get: function () { return flow_1.Flow; } });
tslib_1.__exportStar(require("./lib/flows/flow-helper"), exports);
var flow_run_1 = require("./lib/flow-run/flow-run");
Object.defineProperty(exports, "RunEnvironment", { enumerable: true, get: function () { return flow_run_1.RunEnvironment; } });
var execution_state_1 = require("./lib/flow-run/execution/execution-state");
Object.defineProperty(exports, "ExecutionState", { enumerable: true, get: function () { return execution_state_1.ExecutionState; } });
tslib_1.__exportStar(require("./lib/flows/dto/create-flow-request"), exports);
var id_generator_1 = require("./lib/common/id-generator");
Object.defineProperty(exports, "apId", { enumerable: true, get: function () { return id_generator_1.apId; } });
Object.defineProperty(exports, "ApId", { enumerable: true, get: function () { return id_generator_1.ApId; } });
tslib_1.__exportStar(require("./lib/flows/trigger-events/trigger-events-dto"), exports);
tslib_1.__exportStar(require("./lib/flows/trigger-events/trigger-event"), exports);
tslib_1.__exportStar(require("./lib/flows/sample-data"), exports);
tslib_1.__exportStar(require("./lib/project/update-project-request"), exports);
tslib_1.__exportStar(require("./lib/common/base-model"), exports);
tslib_1.__exportStar(require("./lib/flows/flow-instances"), exports);
tslib_1.__exportStar(require("./lib/flows/folders/folder"), exports);
tslib_1.__exportStar(require("./lib/flows/folders/folder-requests"), exports);
tslib_1.__exportStar(require("./lib/flows/dto/flow-template-request"), exports);
const system_1 = require("@sinclair/typebox/system");
tslib_1.__exportStar(require("./lib/flows/dto/list-flows-request"), exports);
// Look at https://github.com/sinclairzx81/typebox/issues/350
system_1.TypeSystem.ExactOptionalPropertyTypes = false;
