// @activepieces/shared/src/lib/flows/dto/flow-template-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListFlowTemplatesRequest = exports.FlowTemplate = exports.FlowVersionTemplate = void 0;
const typebox_1 = require("@sinclair/typebox");
const flow_version_1 = require("../flow-version");
exports.FlowVersionTemplate = typebox_1.Type.Omit(flow_version_1.FlowVersion, ["id", "created", "updated", "flowId", "state", "updatedBy"]);
exports.FlowTemplate = typebox_1.Type.Object({
    id: typebox_1.Type.String(),
    name: typebox_1.Type.String(),
    description: typebox_1.Type.String(),
    tags: typebox_1.Type.Array(typebox_1.Type.String()),
    pieces: typebox_1.Type.Array(typebox_1.Type.String()),
    pinnedOrder: typebox_1.Type.Union([typebox_1.Type.Null(), typebox_1.Type.Number()]),
    blogUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
    template: exports.FlowVersionTemplate,
});
exports.ListFlowTemplatesRequest = typebox_1.Type.Object({
    pieces: typebox_1.Type.Optional(typebox_1.Type.Array(typebox_1.Type.String())),
    tags: typebox_1.Type.Optional(typebox_1.Type.Array(typebox_1.Type.String())),
    search: typebox_1.Type.Optional(typebox_1.Type.String()),
    pinned: typebox_1.Type.Optional(typebox_1.Type.Boolean())
});
