// @activepieces/shared/src/lib/pieces/dto/piece-requests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstallPieceRequest = exports.PieceOptionRequest = exports.GetPieceRequestQuery = exports.ListPiecesRequestQuery = exports.GetPieceRequestParams = exports.GetPieceRequestWithScopeParams = exports.QueryVerType = exports.SemVerType = void 0;
const typebox_1 = require("@sinclair/typebox");
const semVerRegex = /^[0-9]+\.[0-9]+\.[0-9]+$/;
const queryVersionRegex = /^([~^])?[0-9]+\.[0-9]+\.[0-9]+$/;
exports.SemVerType = typebox_1.Type.RegEx(semVerRegex);
exports.QueryVerType = typebox_1.Type.RegEx(queryVersionRegex);
exports.GetPieceRequestWithScopeParams = typebox_1.Type.Object({
    name: typebox_1.Type.String(),
    scope: typebox_1.Type.String(),
});
exports.GetPieceRequestParams = typebox_1.Type.Object({
    name: typebox_1.Type.String(),
});
exports.ListPiecesRequestQuery = typebox_1.Type.Object({
    release: typebox_1.Type.Optional(exports.SemVerType),
});
exports.GetPieceRequestQuery = typebox_1.Type.Object({
    version: typebox_1.Type.Optional(exports.QueryVerType),
});
exports.PieceOptionRequest = typebox_1.Type.Object({
    pieceVersion: exports.QueryVerType,
    pieceName: typebox_1.Type.String({}),
    stepName: typebox_1.Type.String({}),
    propertyName: typebox_1.Type.String({}),
    input: typebox_1.Type.Any({}),
});
exports.InstallPieceRequest = typebox_1.Type.Object({
    pieceName: typebox_1.Type.String(),
    pieceVersion: exports.SemVerType,
});
